### PR TITLE
Refactor file writing out to decouple from Secrets

### PR DESCRIPTION
**Pull request by mcpherrinm**
This makes a new function which handles atomically writing files with a given
mode and permissions.  I plan to reuse this code in the future, so this is just
the first part of seperation

The package is called "output" because I plan to move the Output* code there
too, but that requires a bit more refactoring to avoid circular dependencies.
So that's another PR for later.

### CODE REVIEW

File: write.go [hunk 1]
```diff
@@ -192,106 +189,30 @@ func (out *OutputDir) Cleanup(secrets map[string]Secret) error {
...
-}
-
-// atomicWrite creates a temporary file, sets perms, writes content, and renames it to filename
-// This sequence ensures the following:
-// 1. Nobody can open the file before we set owner/permissions properly
-// 2. Nobody observes a partially-overwritten secret file.
-// Since keysync is intended to write to tmpfs, this function doesn't do the necessary fsyncs if it
-// were persisting content to disk.
+// Write puts a Secret into OutputDir
 func (out *OutputDir) Write(secret *Secret) (*secretState, error) {
```
Comment (mcpherrinm) [inline write.go]: This function stays the same (and all its tests still pass), just the core file logic is factored out into output.WriteFileAtomically

File: output/write.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
+package output
```
Comment (mcpherrinm) [inline output/write.go]: All of the code in this file is extracted from write.go at the top level: WriteFileAtomically is a new function extracted from OutputDir.Write(Secret), and the rest is straight moved.

File: output/write.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+
+	return &FileInfo{filemode, uid, gid}, nil
+}
+
+// WriteFileAtomically creates a temporary file, sets perms, writes content, and renames it to filename
+// This sequence ensures the following:
+// 1. Nobody can open the file before we set owner/permissions properly
+// 2. Nobody observes a partially-overwritten secret file.
+// The returned FileInfo may not match the passed in one, especially if chownFiles is false.
+func WriteFileAtomically(dir, filename string, chownFiles bool, fileInfo FileInfo, enforceFilesystem Filesystem, content []byte) (*FileInfo, error) {
```
Comment (mcpherrinm) [inline output/write.go]: This function takes a lot of arguments but they're all typed (except the first two) so I think it's okish

File: output/write.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+	}
+
+	return fileinfo, nil
+}
+
+// The Filesystem identification.  On Mac, this is uint32, and int64 on linux
+// So both are safe to store as an int64.
+// Linux Tmpfs = 0x01021994
+// Get these constants with `stat --file-system --format=%t`
+type Filesystem int64
```
Comment (mbyczkowski) [inline output/write.go]: nit: maybe call it `FilesystemID`?

File: write.go [hunk 5]
```diff
@@ -192,125 +189,37 @@ func (out *OutputDir) Cleanup(secrets map[string]Secret) error {
...
+		fileInfo.GID = owner.GID
 	}
 
-	// While this is intended for use with tmpfs, you could write secrets to disk.
-	// We ignore any errors from syncing, as it's not strictly required.
-	_ = f.Sync()
-
-	// Rename is atomic, so nobody will observe a partially updated secret
-	err = os.Rename(fullPath+randSuffix, fullPath)
+	fileinfo, err := output.WriteFileAtomically(out.WriteDirectory, filename, out.ChownFiles, fileInfo, out.EnforceFilesystem, secret.Content)
```
Comment (mbyczkowski) [inline write.go]: this looks much nicer now! 👍

File: output/write.go (hunk 4 above)
Comment (mbyczkowski) [reply output/write.go]: oh, i see this is just a copy. maybe a future improvement/tweak then

File: ownership/lookup.go [hunk 6]
```diff
@@ -10,16 +10,16 @@ import (
 // It is intended to be used with the implementation on Os.  There's also one in mock.go that
 // uses fixed data instead of operating-system sourced data.
 type Lookup interface {
-	UID(username string) (uint32, error)
```
Comment (mbyczkowski) [inline ownership/lookup.go]: Not sure why we're changing it from `uint32` to `int`. uid/gid are unsigned 32bit ints on POSIX systems, IIRC, but I don't know if it matters in reality.

Comment (mcpherrinm) [reply ownership/lookup.go]: ultimately the point of these values are to call Chown, which takes int. 
 https://golang.org/pkg/os/#Chown

File: output/write.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+		return nil, err
+	}
+
+	if enforceFilesystem != 0 {
+		good, err := isFilesystem(f, enforceFilesystem)
+		if err != nil {
+			return nil, fmt.Errorf("checking filesystem type: %v", err)
+		}
+		if !good {
+			return nil, fmt.Errorf("unexpected filesystem writing %s", filename)
```
Comment (jbpeirce) [inline output/write.go]: nit: would logging the expected filesystem here leak important information?  If not, might be useful for debugging.

Comment (mcpherrinm) [reply output/write.go]: This is related to issue #54 where it's hard to understand/debug the filesystem checking.  Added a note about this, there, for future improvement

**Pull request merged**